client/src/pages/files/addTags.tsx

```diff
@@ -0,0 +1,198 @@
+import Prose from 'components/Prose';
+import React, {
+  useState,
+  useEffect,
+  KeyboardEvent,
+  useRef,
+  useMemo,
+} from 'react';
+
+import {getUserFiles, updateDocumentTags} from 'lib/api/files';
+import {UserFile} from 'types/UserFile';
+import {Table} from 'semantic-ui-react';
+
+import {useAuth} from 'contexts/auth';
+
+export default function TaggingView() {
+  const {user} = useAuth();
+  const [userFiles, setUserFiles] = useState<Map<string, [UserFile, boolean]>>(
+    new Map<string, [UserFile, boolean]>()
+  );
+
+  useEffect(() => {
+    if (user) {
+      getUserUploadedFiles();
+    }
+  }, [user]);
+
+  async function getUserUploadedFiles() {
+    const newUserFiles = await getUserFiles(user!);
+    setUserFiles(
+      new Map(
+        Array.from(newUserFiles).map(userFile => {
+          return [userFile.fileName, [userFile, false]];
+        })
+      )
+    );
+  }
+
+  const removeTagFromFile = (tagIndex: number, fileName: string) => {
+    const newFiles = new Map(userFiles);
+    const oldFileInfo = newFiles.get(fileName)!;
+    const newFileInfo: UserFile = {
+      fileName: oldFileInfo[0].fileName,
+      size: oldFileInfo[0].size,
+      contentType: oldFileInfo[0].contentType,
+      tags: oldFileInfo[0].tags.slice(),
+      timeCreated: oldFileInfo[0].timeCreated,
+      userId: oldFileInfo[0].userId,
+    };
+    newFileInfo.tags.splice(tagIndex, 1);
+    newFiles.set(oldFileInfo[0].fileName, [newFileInfo, true]);
+    setUserFiles(newFiles);
+  };
+
+  const addTagToFile = (tag: string, fileName: string) => {
+    if (
+      userFiles
+        .get(fileName)![0]
+        .tags.find(fileTag => fileTag.toLowerCase() === tag.toLowerCase())
+    ) {
+      return false; // False if nothing added
+    }
+    const newFiles = new Map(userFiles);
+    const oldFileInfo = newFiles.get(fileName)!;
+    const newFileInfo: UserFile = {
+      fileName: oldFileInfo[0].fileName,
+      size: oldFileInfo[0].size,
+      contentType: oldFileInfo[0].contentType,
+      tags: [...oldFileInfo[0].tags.slice(), tag],
+      timeCreated: oldFileInfo[0].timeCreated,
+      userId: oldFileInfo[0].userId,
+    };
+    newFiles.set(oldFileInfo[0].fileName, [newFileInfo, true]);
+    setUserFiles(newFiles);
+    return true; // True if a new tag is added
+  };
+
+  const canUploadTags = useMemo(() => !(user === null), [user]);
+
+  const uploadAllTags = () => {
+    if (canUploadTags) {
```

Comment:
Better to say `if (!canUploadTags) return;`, keeping the rest of the function on the outermost scope.